Game/Source/GDKTestGyms/Private/UserExperienceComponent.cpp

```diff
@@ -0,0 +1,184 @@
+#include "UserExperienceComponent.h"
+
+#include "Net/UnrealNetwork.h"
+#include "SpatialNetDriver.h"
+#include "Utils/SpatialMetrics.h"
+
+DEFINE_LOG_CATEGORY(LogUserExperienceComponent);
+
+namespace
+{
+	float Calculate80thPctAverage(const TArray<float>& Array)
+	{
+		TArray<float> Sorted = Array;
+		Sorted.Sort();
+		int ElementsToAverage = Sorted.Num()*0.8f; // Exclude 20% of samples
+
+		float Avg = 0.0f;
+		for (int i = 0; i < ElementsToAverage; i++)
+		{
+			Avg += Sorted[i];
+		}
+		Avg /= ElementsToAverage;
+		return Avg;
+	}
+}
+
+// Sets default values for this component's properties
+UUserExperienceComponent::UUserExperienceComponent()
+{
+	// Set this component to be initialized when the game starts, and to be ticked every frame.  You can turn these features
+	// off to improve performance if you don't need them.
+	PrimaryComponentTick.bCanEverTick = true;
+	bWantsInitializeComponent = true;
+	bAutoActivate = true;
+}
+
+void UUserExperienceComponent::InitializeComponent()
+{
+	ElapsedTime = 0.0f;
+	ServerClientRTT = 0.0f;
+	ServerViewLateness = 0.0f;
+	ClientRTTTimer = 0.0f; 
+	SetIsReplicated(true);
+	UActorComponent::InitializeComponent();
+}
+
+void UUserExperienceComponent::SendServerRPC()
+{
+	ServerRTT(ElapsedTime);
+}
+
+void UUserExperienceComponent::UpdateClientObservations(float DeltaTime)
+{
+	// Prune out any components which went out of view to avoid skewing results
+	TMap<UUserExperienceComponent*, ObservedUpdate> NewObservations;
+	for (TObjectIterator<UUserExperienceComponent> It; It; ++It)
+	{
+		if (this != *It && It->GetOwner() && It->GetOwner()->GetWorld() == GetWorld())
+		{
+			if (ObservedUpdate* PreviousUpdate = ObservedComponents.Find(*It))
+			{
+				// Carry over into new observations
+				NewObservations.Add(*It, MoveTemp(*PreviousUpdate));
+			}
+			else
+			{
+				NewObservations.Add(*It, ObservedUpdate{});
+			}
+		}
+	}
+	ObservedComponents = NewObservations;
+
+	for (TPair<UUserExperienceComponent*, ObservedUpdate>& ObservationPair : ObservedComponents)
+	{
+		ObservedUpdate& Observation = ObservationPair.Value;
+		UUserExperienceComponent* Component = ObservationPair.Key;
+
+		Observation.TimeSinceChange += DeltaTime;
+		if (Observation.Value != Component->ClientTime)
+		{
+			// New value observed.
+			Observation.Value = Component->ClientTime;
+			Observation.TrackedChanges.Push(Observation.TimeSinceChange);
+			Observation.TimeSinceChange = 0.0f;
+
+			if (Observation.TrackedChanges.Num() > NumWindowSamples)
+			{
+				Observation.TrackedChanges.RemoveAt(0);
+			}
+		}
+	}
+}
+
+float UUserExperienceComponent::CalculateWorldFrequency()
+{
+	float Frequency = 0.0f;
+	for (TPair<UUserExperienceComponent*, ObservedUpdate>& ObservationPair : ObservedComponents)
+	{
+		ObservedUpdate& Observation = ObservationPair.Value;
+		UUserExperienceComponent* Component = ObservationPair.Key;
+
+		if (!Component->IsActive())
+		{
+			continue;
+		}
+
+		if (Observation.TrackedChanges.Num() == NumWindowSamples)
```

Comment:
Why this check?>

Reply:
The intention was so that the reported number only starts being populated when enough samples are available to remove outliers